### PR TITLE
Updates the dev server to handle multiple routes matching the same URL

### LINKED ISSUE #3809: 🐛 BUG: `injectRoute` path intercepted by dynamic route during dev
### What version of `astro` are you using?

v1.0.0-beta.73
v1.0.0-beta.63

### Are you using an SSR adapter? If so, which one?

n/a

### What package manager are you using?

npm

### What operating system are you using?

macOS / Stackblitz

### Describe the Bug

Routes added using `injectRoute` are intercepted by matching dynamic routes in a project in dev mode (maybe also in SSR? I haven’t tested that).

For example, say an integration injects a page at `/injected`. If the project also has a `src/pages/[slug].astro` route, the dynamic `[slug].astro` route will try to handle the request to `/injected` and (assuming `getStaticPaths` doesn’t include `slug: 'injected'`) print an error: `Route pattern matched, but no matching static path found`.

The same even applies when you have a `src/pages/[slug]/index.astro` (see second reproduction below), so I’m guessing _any_ root-level dynamic segment would trigger this.

This is an issue with the new image integration, because its `/_image` route breaks in projects using a dynamic route as described above.

Two possibilities:

1. On Discord, Fred suggested this could be a priority issue: we should treat injected routes as higher priority than local ones.

    I don’t know if this is true — what about the scenario where an integration injects a catch-all `[...path]` pattern but a local project can override it? Plausible for some of our speculative “no code, content-first” integration ideas.
    
    We don’t currently mention `injectRoute` in our [route priority docs](https://docs.astro.build/en/core-concepts/routing/#route-priority-order), but my instinct would be that injected routes have a lower priority than a project’s file routes.

2. Similar to #3621, we could treat dynamic routes as a stack where if the first route doesn’t provide a matching static path, the next is evaluated and so on. I personally feel this would be closer to the behaviour I expect.

Potentially related: #3451

#### `src/pages/[slug].astro` intercepts the injected route

https://stackblitz.com/edit/github-zmcb8k?file=astro.config.mjs

#### `src/pages/[slug]/index.astro` intercepts the injected route

https://stackblitz.com/edit/github-zmcb8k-rfjobw

### Participation

- [ ] I am willing to submit a pull request for this issue.

### PR DESCRIPTION
**Pull request by tony-sull**
Closes #3809 

## Changes

- dev server now finds all page templates that match the requested URL and checks each one in priority order
    - in SSR, users will get an error if more than one page template matches a URL in `astro dev`
    - in SSG, the dev server now prioritizes URLs the same as `astro build`

- injected routes are now sorted with a higher priority than all user-provided page routes

## Testing

- all existing `routing-priority` tests pass
- added tests for prioritizing injected routes
- added tests for multiple page routes matching the same URL

## Docs

N/A bug fix only

### CODE REVIEW

**Comment by tony-sull** [comment]
Likely related to #3809 as well, dev routing should prioritize injected routes as well

File: packages/astro/src/core/routing/manifest/create.ts [hunk 1]
```diff
@@ -159,6 +159,29 @@ function comparator(a: Item, b: Item) {
 	return a.file < b.file ? -1 : 1;
 }
 
+function injectedRouteToItem(
```
Comment (tony-sull) [inline packages/astro/src/core/routing/manifest/create.ts:162]: This is used to sort injected routes with the same comparer used for user-defined routes

File: packages/astro/src/core/routing/match.ts [hunk 2]
```diff
@@ -4,3 +4,8 @@ import type { ManifestData, RouteData } from '../../@types/astro';
 export function matchRoute(pathname: string, manifest: ManifestData): RouteData | undefined {
 	return manifest.routes.find((route) => route.pattern.test(pathname));
 }
+
+/** Finds all matching routes from pathname */
+export function matchAllRoutes(pathname: string, manifest: ManifestData): RouteData[] {
```
Comment (tony-sull) [inline packages/astro/src/core/routing/match.ts:9]: used by the dev server to find all potential routes for the requested URL

File: packages/astro/src/vite-plugin-astro-server/index.ts [hunk 3]
```diff
@@ -248,26 +248,81 @@ async function handleRequest(
...
+	matches.map(({ component }) => `- ${component}`).join('\n')
+}
+`);
+		}
+
+		for await (const maybeRoute of matches) {
+			const filePath = new URL(`./${maybeRoute.component}`, config.root);
+			const preloadedComponent = await preload({ astroConfig: config, filePath, viteServer });
+			const [, mod] = preloadedComponent;
+			// attempt to get static paths
```
Comment (matthewp) [inline packages/astro/src/vite-plugin-astro-server/index.ts:267]: does this result in getStaticPaths being called more than once?

Comment (tony-sull) [reply packages/astro/src/vite-plugin-astro-server/index.ts:267]: It's still only called once.  The `getStaticPaths()` result is always cached even if the params don't match

Just noticed we have a test to make sure it's only called once but that test only covers `build`.  I just pushed up a test that checks that for `dev` as well :+1:

**Review by matthewp — APPROVED** [review]

**Pull request merged**